openedx/core/djangoapps/user_api/accounts/tests/test_views.py

```diff
@@ -1129,6 +1055,71 @@ def _get_dead_end_states(self):
+    """
+    Tests the account deactivation/logout endpoint.
+    """
+    def setUp(self):
+        super(TestDeactivateLogout, self).setUp()
+        self.test_password = 'password'
+        self.test_user = UserFactory(password=self.test_password)
+        UserSocialAuth.objects.create(
+            user=self.test_user,
+            provider='some_provider_name',
+            uid='xyz@gmail.com'
+        )
+        UserSocialAuth.objects.create(
+            user=self.test_user,
+            provider='some_other_provider_name',
+            uid='xyz@gmail.com'
+        )
+
+        self.url = reverse('deactivate_logout')
+
+    def build_post(self, password):
+        return {'password': password}
+
+    def test_user_can_deactivate_self(self):
+        """
+        Verify a user calling the deactivation endpoint logs out the user, deletes all their SSO tokens,
+        and creates a user retirement row.
+        """
+        self.client.login(username=self.test_user.username, password=self.test_password)
+        headers = self.build_jwt_headers(self.test_user)
+        response = self.client.post(self.url, self.build_post(self.test_password), **headers)
+        self.assertEqual(response.status_code, status.HTTP_204_NO_CONTENT)
+        updated_user = User.objects.get(id=self.test_user.id)
+        self.assertEqual(get_retired_email_by_email(self.test_user.email), updated_user.email)
+        self.assertFalse(updated_user.has_usable_password())
+        self.assertEqual(list(UserSocialAuth.objects.filter(user=self.test_user)), [])
+        self.assertEqual(len(UserRetirementStatus.objects.filter(user_id=self.test_user.id)), 1)
+
+    def test_password_mismatch(self):
+        """
+        Verify that the user submitting a mismatched password results in
+        a rejection.
+        """
+        self.client.login(username=self.test_user.username, password=self.test_password)
+        headers = self.build_jwt_headers(self.test_user)
+        response = self.client.post(self.url, self.build_post(self.test_password + "xxxx"), **headers)
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+
+    def test_called_twice(self):
+        """
+        Verify a user calling the deactivation endpoint twice results in a "forbidden"
+        error, as the user will be logged out.
+        """
+        self.client.login(username=self.test_user.username, password=self.test_password)
+        headers = self.build_jwt_headers(self.test_user)
+        response = self.client.post(self.url, self.build_post(self.test_password), **headers)
+        self.assertEqual(response.status_code, status.HTTP_204_NO_CONTENT)
+        self.client.login(username=self.test_user.username, password=self.test_password)
+        headers = self.build_jwt_headers(self.test_user)
+        response = self.client.post(self.url, self.build_post(self.test_password), **headers)
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+
+
```

Comment:
I think it's worth to create another test case to make sure after the deactivation endpoint succeeds, the user is logged out, and cannot log back in.

Reply:
I'm going to incorporate this into the top-most test case here, since it already is executing a list of assertions about the user's state after deactivation completes. I agree that it is worth testing that the user cannot log back in.